addons/mail/static/src/models/thread/thread.js

```diff
@@ -956,6 +995,41 @@ function factory(dependencies) {
             return [['link', this.cache()]];
         }
 
+        /**
+         * @private
+         * @returns {integer}
+         */
+        _computeLocalMessageUnreadCounter() {
+            // the only situations where serverMessageUnreadCounter can/have to be
+            // trusted are:
+            // - we have no last message (and then no messages at all)
+            // - the message it used to compute is the last message we know
+            if (this.orderedMessages.length === 0) {
+                return this.serverMessageUnreadCounter;
+            }
+            // from here serverLastMessageId is not undefined because
+            // orderedMessages contain at least one message.
+            if (!this.lastSeenByCurrentPartnerMessageId) {
+                return this.serverMessageUnreadCounter;
+            }
+            const firstMessage = this.orderedMessages[0];
+            // if the lastSeenByCurrentPartnerMessageId is not known (not fetched), then we
+            // need to rely on server value to determine the amount of unread
+            // messages until the last message it knew when computing the
+            // serverMessageUnreadCounter
+            if (this.lastSeenByCurrentPartnerMessageId < firstMessage.id) {
+                const fetchedNotSeenMessages = this.orderedMessages.filter(message =>
+                    message.id > this.serverLastMessageId
+                );
+                return this.serverMessageUnreadCounter + fetchedNotSeenMessages.length;
+            }
+            // lastSeenByCurrentPartnerMessageId is a known message,
+            // then we can forget serverMessageUnreadCounter
+            const maxId = Math.max(this.serverLastMessageId, this.lastSeenByCurrentPartnerMessageId);
+            const fetchedNotSeenMessages = this.orderedMessages.filter(message => message.id > maxId);
+            return fetchedNotSeenMessages.length;
```

Comment:
feels like "reduce" is more appropriate than filter here (it's reducing the list to a single value), though it really won't change the result